universal-application-tool-0.0.1/app/controllers/applicant/ApplicantInformationController.java

```diff
@@ -0,0 +1,115 @@
+import play.mvc.Http;
+import play.mvc.Result;
+import play.mvc.Results;
+import repository.ApplicantRepository;
+import services.applicant.ApplicantData;
+import services.applicant.ApplicantNotFoundException;
+import views.applicant.ApplicantInformationView;
+
+/**
+ * Provides methods for editing and updating an applicant's information, such as their preferred
+ * language.
+ */
+public final class ApplicantInformationController extends CiviFormController {
+
+  private final HttpExecutionContext httpExecutionContext;
+  private final MessagesApi messagesApi;
+  private final ApplicantInformationView informationView;
+  private final ApplicantRepository repository;
+  private final FormFactory formFactory;
+  private final ProfileUtils profileUtils;
+
+  @Inject
+  public ApplicantInformationController(
+      HttpExecutionContext httpExecutionContext,
+      MessagesApi messagesApi,
+      ApplicantInformationView informationView,
+      ApplicantRepository repository,
+      FormFactory formFactory,
+      ProfileUtils profileUtils) {
+    this.httpExecutionContext = httpExecutionContext;
+    this.messagesApi = messagesApi;
+    this.informationView = informationView;
+    this.repository = repository;
+    this.formFactory = formFactory;
+    this.profileUtils = profileUtils;
+  }
+
+  public CompletionStage<Result> edit(Http.Request request, long applicantId) {
+    return checkApplicantAuthorization(profileUtils, request, applicantId)
+        .thenApplyAsync(
+            v -> ok(informationView.render(request, applicantId)), httpExecutionContext.current())
+        .exceptionally(
+            ex -> {
+              if (ex instanceof CompletionException) {
+                if (ex.getCause() instanceof SecurityException) {
+                  return unauthorized();
+                }
+              }
+              throw new RuntimeException(ex);
+            });
+  }
+
+  public CompletionStage<Result> update(Http.Request request, long applicantId) {
+    Form<ApplicantInformationForm> form = formFactory.form(ApplicantInformationForm.class);
+    if (form.hasErrors()) {
+      return supplyAsync(Results::badRequest);
+    }
+    ApplicantInformationForm infoForm = form.bindFromRequest(request).get();
+
+    return checkApplicantAuthorization(profileUtils, request, applicantId)
+        .thenComposeAsync(
+            v -> repository.lookupApplicant(applicantId), httpExecutionContext.current())
+        .thenComposeAsync(
+            maybeApplicant -> {
+              // Set preferred locale.
+              if (maybeApplicant.isPresent()) {
+                Applicant applicant = maybeApplicant.get();
+                ApplicantData data = applicant.getApplicantData();
+                data.setPreferredLocale(infoForm.getLocale());
+                return repository
+                    .updateApplicant(applicant)
+                    .thenApplyAsync(v -> applicant, httpExecutionContext.current());
```

Comment:
question: What does this line do? Does this just make sure we return `applicant` after the update happens?

Reply:
Yup! We need the applicant entity in the next `thenComposeAsync`, so we have to return it explicitly here

Reply:
I added a comment so this is clearer